plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java

```diff
@@ -2239,211 +2305,309 @@ public LibvirtVMDef createVMFromSpec(final VirtualMachineTO vmTO) {
+        if (vmTO.getMinRam() != vmTO.getMaxRam() && !_noMemBalloon) {
+            grd.setMemBalloning(true);
+            grd.setCurrentMem(vmTO.getMinRam() / 1024);
+        }
+        grd.setVcpuNum(vmTO.getCpus());
+        return grd;
+    }
 
-        final ConsoleDef console = new ConsoleDef("pty", null, null, (short)0);
-        devices.addDevice(console);
+    private void configureGuestIfUefiEnabled(boolean isSecureBoot, String bootMode, GuestDef guest) {
+        setGuestLoader(bootMode, SECURE, guest, GuestDef.GUEST_LOADER_SECURE);
+        setGuestLoader(bootMode, LEGACY, guest, GuestDef.GUEST_LOADER_LEGACY);
 
-        //add the VNC port passwd here, get the passwd from the vmInstance.
-        final String passwd = vmTO.getVncPassword();
-        final GraphicDef grap = new GraphicDef("vnc", (short)0, true, vmTO.getVncAddr(), passwd, null);
-        devices.addDevice(grap);
+        if (isUefiPropertieNotNull(GuestDef.GUEST_NVRAM_PATH)) {
+            guest.setNvram(_uefiProperties.getProperty(GuestDef.GUEST_NVRAM_PATH));
+        }
 
-        final InputDef input = new InputDef("tablet", "usb");
-        devices.addDevice(input);
+        if (isSecureBoot && isUefiPropertieNotNull(GuestDef.GUEST_NVRAM_TEMPLATE_SECURE) && SECURE.equalsIgnoreCase(bootMode)) {
+            guest.setNvramTemplate(_uefiProperties.getProperty(GuestDef.GUEST_NVRAM_TEMPLATE_SECURE));
+        } else if (isUefiPropertieNotNull(GuestDef.GUEST_NVRAM_TEMPLATE_LEGACY)) {
+            guest.setNvramTemplate(_uefiProperties.getProperty(GuestDef.GUEST_NVRAM_TEMPLATE_LEGACY));
+        }
+    }
 
-        // Add an explicit USB devices for ARM64
-        if (_guestCpuArch != null && _guestCpuArch.equals("aarch64")) {
-            devices.addDevice(new InputDef("keyboard", "usb"));
-            devices.addDevice(new InputDef("mouse", "usb"));
-            devices.addDevice(new LibvirtVMDef.USBDef((short)0, 0, 5, 0, 0));
+    private void setGuestLoader(String bootMode, String mode, GuestDef guest, String propertie) {
+        if (isUefiPropertieNotNull(propertie) && mode.equalsIgnoreCase(bootMode)) {
+            guest.setLoader(_uefiProperties.getProperty(propertie));
         }
+    }
 
-        DiskDef.DiskBus busT = getDiskModelFromVMDetail(vmTO);
+    private boolean isUefiPropertieNotNull(String propertie) {
+        return _uefiProperties.getProperty(propertie) != null;
+    }
 
-        if (busT == null) {
-            busT = getGuestDiskModel(vmTO.getPlatformEmulator(), isUefiEnabled);
-        }
+    private boolean isGuestAarch64() {
+        return AARCH64.equals(_guestCpuArch);
+    }
 
-        // If we're using virtio scsi, then we need to add a virtual scsi controller
-        if (busT == DiskDef.DiskBus.SCSI) {
-            final SCSIDef sd = new SCSIDef((short)0, 0, 0, 9, 0, vcpus);
-            devices.addDevice(sd);
-        }
+    /**
+     * Create a guest definition from a VM specification.
```

Comment:
Some documentations have the verb as Imperative (e.g. **create**), some Indicative (e.g. it **creates**).
This is a small observation, but maybe you can normalize them. (I would go to the indicative form; but feel free to choose)